src/bin/ion/commands/stats.rs

```diff
@@ -117,32 +109,24 @@ fn analyze_data_stream<Input: IonInput>(reader: &mut SystemReader<AnyEncoding, I
     return out;
 }
 
-fn top_level_max_depth(value: LazyValue<AnyEncoding>, depth: usize) -> usize {
-    if value.is_container() {
-        if value.ion_type() == IonType::Struct {
-            value
-                .read()
-                .unwrap()
-                .expect_struct()
-                .unwrap()
-                .iter()
-                .map(|v| crate::commands::stats::top_level_max_depth(v.unwrap().value(), depth + 1))
-                .max()
-                .unwrap()
-        } else {
-            value
-                .read()
-                .unwrap()
-                .expect_list()
-                .unwrap()
-                .iter()
-                .map(|v| top_level_max_depth(v.unwrap(), depth + 1))
-                .max()
-                .unwrap()
+fn top_level_max_depth(value: LazyValue<AnyEncoding>) -> usize {
+    let mut max_depth = 0;
+    let mut stack = vec![(value, 0)];
+    while let Some((current_value, depth)) = stack.pop() {
+        max_depth = max(max_depth, depth);
+        if current_value.ion_type() == IonType::Struct {
+            let struct_value = current_value.read().unwrap().expect_struct().unwrap();
+            for field in struct_value {
+                stack.push((field.unwrap().value(), depth + 1));
+            }
+        } else if current_value.ion_type() == IonType::List {
```

Comment:
This branch should include s-expressions too, right?

Reply:
Thanks for calling it out, `s-expressions` should be processed too.  In my opinion, there should be another branch to process `s-expressions` specifically. Please correct me if I'm wrong, but as I understand it, the current way we read the value out of `LazyValue` doesn't allow us to merge both `lists` and `s-expressions` into one branch. We can only call `current_value.read().unwrap().expect_list() `and `current_value.read().unwrap().expect_sexp()` separately.

Reply:
Sure, that's fine.